File: Rude-Chat-flask-agents/MCP-Server/MCPServer/Services/Agents/DocumentsAgent.cs
using MCPServer.Models;
using MCPServer.Services.Azure;

namespace MCPServer.Services.Agents;

/// <summary>
/// Specialized agent for document management operations.
/// Handles document upload, processing, search, and retrieval.
/// </summary>
public class DocumentsAgent : BaseAgent
{
    private readonly IAzureDocumentService _documentService;

    public DocumentsAgent(
        ILogger<DocumentsAgent> logger, 
        IConfiguration configuration,
        IAzureDocumentService documentService) 
        : base(logger, configuration)
    {
        _documentService = documentService;
    }

    public override string AgentId => "documents-agent";
    public override string Name => "Documents Agent";
    public override string Description => "Specialized agent for document management including upload, processing, search, and retrieval";
    public override IEnumerable<string> Domains => new[] { "documents", "files", "upload", "search", "rag", "content" };

    protected override Task OnInitializeAsync(string? userToken)
    {
        _logger.LogInformation("Initializing Documents Agent");
        // Document service doesn't need user token initialization currently
        return Task.CompletedTask;
    }

    public override Task<IEnumerable<McpTool>> GetAvailableToolsAsync()
    {        try
        {
            // Document operations are typically handled through dedicated endpoints,
            // not as MCP tools, but we can provide metadata about capabilities
            var tools = new List<McpTool>
            {
                new McpTool
                {
                    Name = "search_documents",
                    Description = "Search through uploaded documents using AI search capabilities",
                    InputSchema = new McpToolInputSchema
                    {
                        Type = "object",
                        Properties = new Dictionary<string, McpProperty>
                        {
                            ["query"] = new McpProperty
                            {
                                Type = "string",
                                Description = "The search query to find relevant documents"
                            },
                            ["maxResults"] = new McpProperty
                            {
                                Type = "integer",
                                Description = "Maximum number of results to return (default: 5)"
                            }
                        },
                        Required = new[] { "query" }
                    }
                },
                new McpTool
                {
                    Name = "list_user_documents",
                    Description = "List all documents uploaded by the current user",
                    InputSchema = new McpToolInputSchema
                    {
                        Type = "object",
                        Properties = new Dictionary<string, McpProperty>(),
                        Required = Array.Empty<string>()
                    }
                },
                new McpTool
                {
                    Name = "get_document_content",
                    Description = "Get the full extracted text content of a specific document by its ID",
                    InputSchema = new McpToolInputSchema
                    {
                        Type = "object",
                        Properties = new Dictionary<string, McpProperty>
                        {
                            ["documentId"] = new McpProperty
                            {
                                Type = "string",
                                Description = "The ID of the document to retrieve content for"
                            }
                        },
                        Required = new[] { "documentId" }
                    }
                }
            };

            _logger.LogDebug("Documents Agent providing {ToolCount} tools", tools.Count);
            return Task.FromResult<IEnumerable<McpTool>>(tools);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to get Documents tools");
            return Task.FromResult(Enumerable.Empty<McpTool>());
        }
    }

    public override async Task<McpToolCallResponse> ExecuteToolAsync(McpToolCallRequest request)
    {
        _logger.LogInformation("Documents Agent executing tool: {ToolName}", request.Name);

        try
        {            return request.Name switch
            {
                "search_documents" => await ExecuteSearchDocumentsAsync(request.Arguments),
                "list_user_documents" => await ExecuteListUserDocumentsAsync(request.Arguments),
                "get_document_content" => await ExecuteGetDocumentContentAsync(request.Arguments),
                _ => CreateErrorResponse($"Unknown document tool: {request.Name}")
            };
        }
        catch (Exception ex)
        {
            return CreateErrorResponse($"Documents Agent failed to execute tool {request.Name}", ex);
        }
    }

    public override Task<bool> CanHandleToolAsync(string toolName)
    {
        return Task.FromResult(IsDocumentTool(toolName));
    }    protected override Task<bool> PerformHealthCheckAsync()
    {
        try
        {
            // Check if required services are configured
            var storageConnectionString = _configuration.GetConnectionString("AzureStorage");
            var searchEndpoint = _configuration["AzureAISearch:Endpoint"];
            
            if (string.IsNullOrEmpty(storageConnectionString) || storageConnectionString.Contains("[") ||
                string.IsNullOrEmpty(searchEndpoint) || searchEndpoint.Contains("["))
            {
                return Task.FromResult(false);
            }

            return Task.FromResult(true);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Documents Agent health check failed");
            return Task.FromResult(false);
        }
    }

    protected override Task<Dictionary<string, object>> GetHealthMetadataAsync()
    {
        var metadata = new Dictionary<string, object>
        {
            ["service_type"] = "Document Management",
            ["storage_configured"] = !string.IsNullOrEmpty(_configuration.GetConnectionString("AzureStorage")) && 
                                   !_configuration.GetConnectionString("AzureStorage")!.Contains("["),
            ["search_configured"] = !string.IsNullOrEmpty(_configuration["AzureAISearch:Endpoint"]) && 
                                  !_configuration["AzureAISearch:Endpoint"]!.Contains("["),
            ["openai_configured"] = !string.IsNullOrEmpty(_configuration["AzureOpenAI:Endpoint"]) && 
                                  !_configuration["AzureOpenAI:Endpoint"]!.Contains("[")
        };

        return Task.FromResult(metadata);
    }    private async Task<McpToolCallResponse> ExecuteSearchDocumentsAsync(Dictionary<string, object> arguments)
    {
        var query = arguments.GetValueOrDefault("query")?.ToString();
        var maxResults = int.TryParse(arguments.GetValueOrDefault("maxResults")?.ToString(), out var max) ? max : 5;
        
        // Get userId and sessionId from the arguments (provided by AgentOrchestrator)
        var userId = arguments.GetValueOrDefault("userId")?.ToString();
        var sessionId = arguments.GetValueOrDefault("sessionId")?.ToString();

        if (string.IsNullOrEmpty(query))
        {
            return CreateErrorResponse("Search query is required");
        }
        
        if (string.IsNullOrEmpty(userId))
        {
            return CreateErrorResponse("User ID is required but not provided in the context");
        }
        
        if (string.IsNullOrEmpty(sessionId))
        {
            return CreateErrorResponse("Session ID is required but not provided in the context");
        }

        try
        {
            var results = await _documentService.SearchDocumentsAsync(query, userId, sessionId, maxResults);
            
            if (!results.Any())
            {
                return CreateSuccessResponse($"No documents found matching query: '{query}'");
            }

            var responseText = $"📄 **Document Search Results** (Query: '{query}')\n\n";
            foreach (var (result, index) in results.Select((r, i) => (r, i + 1)))
            {
                responseText += $"**{index}. {result.FileName}**\n";
                responseText += $"• Score: {result.Score:F2}\n";
                var contentLength = result.Content?.Length ?? 0;
                var previewLength = Math.Min(200, contentLength);
                responseText += $"• Content: {result.Content?[..previewLength]}...\n";
                responseText += $"• Uploaded: {result.UploadedAt:yyyy-MM-dd HH:mm}\n\n";
            }

            return CreateSuccessResponse(responseText);
        }
        catch (Exception ex)
        {
            return CreateErrorResponse($"Error searching documents: {ex.Message}");
        }
    }    private async Task<McpToolCallResponse> ExecuteListUserDocumentsAsync(Dictionary<string, object> arguments)
    {
        // Get userId and sessionId from the arguments (provided by AgentOrchestrator)
        var userId = arguments.GetValueOrDefault("userId")?.ToString();
        var sessionId = arguments.GetValueOrDefault("sessionId")?.ToString();
        
        if (string.IsNullOrEmpty(userId))
        {
            return CreateErrorResponse("User ID is required but not provided in the context");
        }

        try
        {
            var documents = await _documentService.GetUserDocumentsAsync(userId, sessionId);
            
            if (!documents.Any())
            {
                return CreateSuccessResponse("No documents found for the current user.");
            }

            var responseText = $"📄 **User Documents** ({documents.Count()} found)\n\n";
            foreach (var (doc, index) in documents.Select((d, i) => (d, i + 1)))
            {
                responseText += $"**{index}. {doc.FileName}**\n";
                responseText += $"• ID: {doc.DocumentId}\n";
                responseText += $"• Size: {doc.FileSize:N0} bytes\n";
                responseText += $"• Uploaded: {doc.UploadDate:yyyy-MM-dd HH:mm}\n";
                responseText += $"• Status: {(DocumentStatus)doc.Status}\n\n";
            }            return CreateSuccessResponse(responseText);
        }
        catch (Exception ex)
        {
            return CreateErrorResponse($"Error listing user documents: {ex.Message}");
        }
    }

    private async Task<McpToolCallResponse> ExecuteGetDocumentContentAsync(Dictionary<string, object> arguments)
    {
        var documentId = arguments.GetValueOrDefault("documentId")?.ToString();
        var userId = arguments.GetValueOrDefault("userId")?.ToString();

        if (string.IsNullOrEmpty(documentId))
        {
            return CreateErrorResponse("Document ID is required");
        }
        
        if (string.IsNullOrEmpty(userId))
        {
            return CreateErrorResponse("User ID is required but not provided in the context");
        }

        try
        {
            var content = await _documentService.GetDocumentContentAsync(documentId, userId);
            
            if (string.IsNullOrEmpty(content))
            {
                return CreateSuccessResponse($"Document {documentId} was found but contains no extractable text content.");
            }

            var responseText = $"📄 **Document Content** (ID: {documentId})\n\n{content}";
            return CreateSuccessResponse(responseText);
        }
        catch (Exception ex)
        {
            return CreateErrorResponse($"Error getting document content: {ex.Message}");
        }
    }    private static bool IsDocumentTool(string toolName)
    {
        return toolName switch
        {
            "search_documents" => true,
            "list_user_documents" => true,
            "get_document_content" => true,
            _ => false
        };
    }

    /// <summary>
    /// Documents-specific method to suggest document operations
    /// </summary>
    public Task<McpToolCallResponse> SuggestDocumentOperationsAsync(string context)
    {
        try
        {
            _logger.LogInformation("Documents Agent providing suggestions for context: {Context}", context);

            var suggestions = new List<string>();

            if (context.ToLower().Contains("search") || context.ToLower().Contains("find"))
            {
                suggestions.Add("Use: search_documents to find relevant content in uploaded documents");
                suggestions.Add("Example: search_documents with query 'contract terms'");
            }

            if (context.ToLower().Contains("list") || context.ToLower().Contains("show"))
            {
                suggestions.Add("Use: list_user_documents to see all your uploaded documents");
            }

            if (context.ToLower().Contains("upload"))
            {
                suggestions.Add("Use the /api/documents/upload endpoint to upload new documents");
                suggestions.Add("Documents are automatically processed and indexed for search");
            }

            if (!suggestions.Any())
            {
                suggestions.Add("Available operations:");
                suggestions.Add("• search_documents - Find content in uploaded documents");
                suggestions.Add("• list_user_documents - View all your documents");
                suggestions.Add("• Upload via API endpoint for new documents");
            }

            var suggestionText = "📄 **Documents Agent Suggestions:**\n\n" + 
                               string.Join("\n", suggestions.Select((s, i) => $"{i + 1}. {s}"));

            return Task.FromResult(CreateSuccessResponse(suggestionText));
        }
        catch (Exception ex)
        {
            return Task.FromResult(CreateErrorResponse("Failed to generate Documents suggestions", ex));
        }
    }
}

File: Rude-Chat-flask-agents/Python-MCP-Server/src/agents/documents_agent.py
"""
Documents agent for document management operations.
Handles document upload, processing, search, and retrieval.
"""

import logging
from typing import List, Dict, Any, Optional

from .base_agent import BaseAgent
try:
    from ..models.mcp_models import McpTool, McpToolCallRequest, McpToolCallResponse, McpProperty, McpToolInputSchema
except ImportError:
    # Fallback for when running directly
    import sys
    import os
    sys.path.append(os.path.dirname(os.path.dirname(__file__)))
    from models.mcp_models import McpTool, McpToolCallRequest, McpToolCallResponse, McpProperty, McpToolInputSchema

logger = logging.getLogger(__name__)

class DocumentsAgent(BaseAgent):
    """Specialized agent for document management operations"""
    
    def __init__(self, config: Dict[str, Any]):
        super().__init__(config)
        self.document_service = None
        
    @property
    def agent_id(self) -> str:
        return "documents-agent"
    
    @property
    def name(self) -> str:
        return "Documents Agent"
    
    @property
    def description(self) -> str:
        return "Specialized agent for document management including upload, processing, search, and retrieval"
    
    @property
    def domains(self) -> List[str]:
        return ["documents", "files", "upload", "search", "rag", "content"]
    
    async def _on_initialize_async(self, user_token: Optional[str]) -> None:
        """Initialize the Documents Agent"""
        logger.info("🔧 Initializing Documents Agent")
        try:
            # Import AzureDocumentService
            from ..services.azure_document_service import AzureDocumentService
            
            # Get cosmos sessions container if available
            cosmos_container = getattr(self, '_cosmos_sessions_container', None)
            logger.info(f"🗄️ Cosmos container available: {cosmos_container is not None}")
            
            # Initialize Azure Document Service
            self.document_service = AzureDocumentService(self._config, cosmos_container)
            logger.info("✅ Documents Agent initialized with AzureDocumentService")
            
            # Test document service initialization
            if self.document_service:
                logger.info("✅ Document service is available and ready")
            else:
                logger.error("❌ Document service failed to initialize")
                
        except Exception as e:
            logger.error(f"❌ Failed to initialize Documents Agent: {e}", exc_info=True)
            self.document_service = None
        # TODO: Initialize Azure Document Intelligence and Search services
    
    async def get_available_tools_async(self) -> List[McpTool]:
        """Get all tools that this agent can execute"""
        try:
            tools = [
                McpTool(
                    name="search_documents",
                    description="Search through uploaded documents using AI search capabilities",
                    input_schema=McpToolInputSchema(
                        type="object",
                        properties={
                            "query": McpProperty(
                                type="string",
                                description="The search query to find relevant documents"
                            ),
                            "maxResults": McpProperty(
                                type="integer",
                                description="Maximum number of results to return (default: 5)"
                            )
                        },
                        required=["query"]
                    )
                ),
                McpTool(
                    name="list_user_documents",
                    description="List all documents uploaded by the current user",
                    input_schema=McpToolInputSchema(
                        type="object",
                        properties={},
                        required=[]
                    )
                ),
                McpTool(
                    name="get_document_content",
                    description="Get the full extracted text content of a specific document by its ID",
                    input_schema=McpToolInputSchema(
                        type="object",
                        properties={
                            "documentId": McpProperty(
                                type="string",
                                description="The ID of the document to retrieve content for"
                            )
                        },
                        required=["documentId"]
                    )
                )
            ]
            
            logger.debug(f"Documents Agent providing {len(tools)} tools")
            return tools
        except Exception as e:
            logger.error(f"Failed to get Documents tools: {str(e)}")
            return []
    
    async def execute_tool_async(self, request: McpToolCallRequest) -> McpToolCallResponse:
        """Execute a tool request"""
        logger.info(f"Documents Agent executing tool: {request.name}")
        
        try:
            if request.name == "search_documents":
                return await self._execute_search_documents(request.arguments)
            elif request.name == "list_user_documents":
                return await self._execute_list_user_documents(request.arguments)
            elif request.name == "get_document_content":
                return await self._execute_get_document_content(request.arguments)
            else:
                return self._create_cannot_answer_response(f"Unknown documents tool: {request.name}")
        except Exception as e:
            return self._create_error_response(f"Documents Agent failed to execute tool {request.name}", e)
    
    async def can_handle_tool_async(self, tool_name: str) -> bool:
        """Check if this agent can handle the specified tool"""
        document_tools = ["search_documents", "list_user_documents", "get_document_content"]
        return tool_name in document_tools
    
    async def _perform_health_check_async(self) -> bool:
        """Perform Documents Agent health check"""
        try:
            # Check if Azure Search and Document Intelligence are configured
            search_endpoint = self._config.get("Azure", {}).get("Search", {}).get("Endpoint")
            doc_intel_endpoint = self._config.get("Azure", {}).get("DocumentIntelligence", {}).get("Endpoint")
            
            if not search_endpoint or not doc_intel_endpoint:
                logger.warning("Documents Agent: Azure Search or Document Intelligence not configured")
                return False
                
            # Simple health check - verify we can get tools
            tools = await self.get_available_tools_async()
            return len(tools) > 0
        except Exception as e:
            logger.warning(f"Documents Agent health check failed: {str(e)}")
            return False
    
    async def _execute_search_documents(self, arguments: Dict[str, Any]) -> McpToolCallResponse:
        """Execute the search_documents tool"""
        # Extract userId and sessionId BEFORE handling nested kwargs from Semantic Kernel
        user_id = arguments.get("userId", "") or arguments.get("user_id", "")
        session_id = arguments.get("sessionId", "") or arguments.get("session_id", "")
        
        # Handle nested kwargs from Semantic Kernel
        if "kwargs" in arguments:
            kwargs_data = arguments["kwargs"]
            # If userId/sessionId are not found at top level, check inside kwargs
            if not user_id:
                user_id = kwargs_data.get("userId", "") or kwargs_data.get("user_id", "")
            if not session_id:
                session_id = kwargs_data.get("sessionId", "") or kwargs_data.get("session_id", "")
            arguments = kwargs_data
        
        query = arguments.get("query", "")
        max_results = arguments.get("maxResults", 5) or arguments.get("max_results", 5)
        
        logger.info(f"🔍 Executing search_documents: query='{query}', user_id='{user_id}', max_results={max_results}")
        
        if not query:
            return self._create_error_response("No search query provided")
        
        if not user_id:
            return self._create_error_response("No user ID provided")
        
        try:
            if not self.document_service:
                logger.error("❌ Document service not available during search")
                return self._create_error_response("Document service not available")
            
            logger.info("📋 Document service is available, performing search...")
            
            # Perform search using AzureDocumentService
            results = await self.document_service.search_documents_async(
                query=query,
                user_id=user_id,
                session_id=session_id,
                max_results=max_results
            )
            
            logger.info(f"🔍 Search completed, found {len(results) if results else 0} results")
            
            if not results:
                return McpToolCallResponse.success(f"No documents found matching '{query}'")
            
            # Format results
            message = f"Found {len(results)} documents matching '{query}':\n\n"
            for i, chunk in enumerate(results, 1):
                message += f"**{i}. {chunk.file_name}** (Chunk {chunk.chunk_index})\n"
                message += f"Content: {chunk.content[:200]}...\n\n"
            
            logger.info(f"✅ Search results formatted successfully")
            return McpToolCallResponse.success(message)
        
        except Exception as e:
            logger.error(f"❌ Error searching documents: {e}", exc_info=True)
            return self._create_error_response(f"Document search failed: {str(e)}")
    
    async def _execute_list_user_documents(self, arguments: Dict[str, Any]) -> McpToolCallResponse:
        """Execute the list_user_documents tool"""
        # Extract userId and sessionId BEFORE handling nested kwargs from Semantic Kernel
        user_id = arguments.get("userId", "") or arguments.get("user_id", "")
        session_id = arguments.get("sessionId", "") or arguments.get("session_id", "")
        
        # Handle nested kwargs from Semantic Kernel
        if "kwargs" in arguments:
            kwargs_data = arguments["kwargs"]
            # If userId/sessionId are not found at top level, check inside kwargs
            if not user_id:
                user_id = kwargs_data.get("userId", "") or kwargs_data.get("user_id", "")
            if not session_id:
                session_id = kwargs_data.get("sessionId", "") or kwargs_data.get("session_id", "")
            arguments = kwargs_data
        
        if not user_id:
            return self._create_error_response("No user ID provided")
        
        try:
            if not self.document_service:
                return self._create_error_response("Document service not available")
            
            # Get user documents using AzureDocumentService
            documents = await self.document_service.get_user_documents_async(
                user_id=user_id,
                session_id=session_id
            )
            
            if not documents:
                return McpToolCallResponse.success("No documents found for this user/session.")
            
            # Format results
            message = f"Found {len(documents)} documents:\n\n"
            for i, doc in enumerate(documents, 1):
                status_text = {0: "Uploaded", 1: "Processing", 2: "Processed", 5: "Failed"}.get(doc.status, "Unknown")
                message += f"**{i}. {doc.file_name}**\n"
                message += f"   - ID: {doc.document_id}\n"
                message += f"   - Status: {status_text}\n"
                message += f"   - Uploaded: {doc.upload_date}\n"
                message += f"   - Size: {doc.file_size} bytes\n\n"
            
            return McpToolCallResponse.success(message)
        
        except Exception as e:
            logger.error(f"Error listing user documents: {e}")
            return self._create_error_response(f"Document listing failed: {str(e)}")
    
    async def _execute_get_document_content(self, arguments: Dict[str, Any]) -> McpToolCallResponse:
        """Execute the get_document_content tool"""
        # Extract userId, sessionId, and documentId BEFORE handling nested kwargs from Semantic Kernel
        user_id = arguments.get("userId", "") or arguments.get("user_id", "")
        session_id = arguments.get("sessionId", "") or arguments.get("session_id", "")
        document_id = (arguments.get("documentId", "") or 
                      arguments.get("document_id", "") or 
                      arguments.get("id", ""))  # SK sometimes sends 'id' instead of 'documentId'
        
        # Handle nested kwargs from Semantic Kernel
        if "kwargs" in arguments:
            kwargs_data = arguments["kwargs"]
            # If userId/sessionId/documentId are not found at top level, check inside kwargs
            if not user_id:
                user_id = kwargs_data.get("userId", "") or kwargs_data.get("user_id", "")
            if not session_id:
                session_id = kwargs_data.get("sessionId", "") or kwargs_data.get("session_id", "")
            if not document_id:
                document_id = (kwargs_data.get("documentId", "") or 
                              kwargs_data.get("document_id", "") or 
                              kwargs_data.get("id", ""))  # SK sometimes sends 'id' instead of 'documentId'
            arguments = kwargs_data
        
        logger.info(f"🔍 Executing get_document_content: document_id='{document_id}', user_id='{user_id}', session_id='{session_id}'")
        
        if not document_id:
            return self._create_error_response("No document ID provided")
        
        if not user_id:
            return self._create_error_response("No user ID provided")
        
        # IMPORTANT: Always require sessionId for security - only allow access to documents in current session
        if not session_id:
            return self._create_error_response("No session ID provided - session isolation required")
        
        try:
            if not self.document_service:
                return self._create_error_response("Document service not available")
            
            # Find document context first with session verification
            document_context = await self.document_service._find_document_with_context_async(document_id, session_id)
            if not document_context:
                return self._create_error_response(f"Document not found: {document_id}")
            
            document_metadata, doc_user_id, doc_session_id = document_context
            
            # Verify the document belongs to the current user and session
            if doc_user_id != user_id:
                logger.warning(f"🚫 Access denied: Document {document_id} belongs to user {doc_user_id}, requested by {user_id}")
                return self._create_error_response("Access denied: Document not found")
            
            if doc_session_id != session_id:
                logger.warning(f"🚫 Session isolation: Document {document_id} belongs to session {doc_session_id}, requested from session {session_id}")
                return self._create_error_response("Document not found in current session")
            
            logger.info(f"✅ Document access authorized: {document_id} for user {user_id} in session {session_id}")
            
            # Extract text
            blob_url = document_metadata.get('blobUrl', '')
            file_name = document_metadata.get('fileName', 'Unknown')
            
            if not blob_url:
                return self._create_error_response("Document content not available")
            
            extracted_text = await self.document_service._extract_text_async(blob_url, file_name)
            
            if not extracted_text:
                return self._create_error_response("Could not extract text from document")
            
            # Format response
            message = f"**Content of {file_name}:**\n\n{extracted_text}"
            
            return McpToolCallResponse.success(message)
        
        except Exception as e:
            logger.error(f"Error getting document content: {e}")
            return self._create_error_response(f"Document content retrieval failed: {str(e)}")
